File: 1340-jump-game-v/1340-jump-game-v.js
/**
 * @param {number[]} arr
 * @param {number} d
 * @return {number}
 */
var maxJumps = function(arr, d) {
    const n=arr.length;
    const dp=new Array(n).fill(-1);
    const dfs=(i)=> {
        if(dp[i] !== -1) {
            return dp[i];
        } 
        dp[i]=1;
        for(let j=i-1; j>=0 && i-j <= d && arr[i] > arr[j]; j--) {
            dfs(j);
            dp[i] = Math.max(dp[i], dp[j]+1);
        }
        for(let j=i+1; j<n && j-i <= d && arr[i] > arr[j]; j++) {
            dfs(j);
            dp[i] = Math.max(dp[i], dp[j]+1);
        }
    };
    for(let i=0;i<n;i++){
        dfs(i);
    }
    return Math.max(...dp);
};

File: 1340-jump-game-v/1340-jump-game-v.cpp
class Solution {
public:
    vector<int>dp;
    int n,d;
    int dfs(int i, vector<int>&a) {
        if(dp[i] != -1) {
            return dp[i];
        }
        int ans=1;
        for(int j=i+1; j<=min(n-1, i+d); j++) { // to right
            if(a[j] >= a[i]) break;
            ans=max(ans,1+dfs(j,a));
        }
        for (int j=i-1; j>= max(0, i-d); j--) { // to left
            if(a[j] >= a[i]) break;
            ans=max(ans,1+dfs(j,a));
        }
        return dp[i]=ans;
    }
    
    int maxJumps(vector<int>& arr, int D) {
        n=arr.size(), d=D;
        dp.assign(n,-1);
        int ans=1;
        for(int i=0;i<n;i++){
            ans=max(ans,dfs(i,arr));
        }
        return ans;
    }
};